// @motify/components/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  MotiView: true,
  View: true,
  MotiText: true,
  Text: true,
  MotiImage: true,
  Image: true,
  MotiScrollView: true,
  ScrollView: true,
  MotiSafeAreaView: true,
  SafeAreaView: true
};
exports.SafeAreaView = exports.MotiSafeAreaView = exports.ScrollView = exports.MotiScrollView = exports.Image = exports.MotiImage = exports.Text = exports.MotiText = exports.View = exports.MotiView = void 0;

var _core = require("@motify/core");

var _reactNative = require("react-native");

var _progress = require("./progress");

Object.keys(_progress).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _progress[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _progress[key];
    }
  });
});
const View = (0, _core.motify)(_reactNative.View)();
exports.View = exports.MotiView = View;
const Text = (0, _core.motify)(_reactNative.Text)();
exports.Text = exports.MotiText = Text;
const Image = (0, _core.motify)(_reactNative.Image)();
exports.Image = exports.MotiImage = Image;
const ScrollView = (0, _core.motify)(_reactNative.ScrollView)();
exports.ScrollView = exports.MotiScrollView = ScrollView;
const SafeAreaView = (0, _core.motify)(_reactNative.SafeAreaView)();
exports.SafeAreaView = exports.MotiSafeAreaView = SafeAreaView;
